ENH: Add a numeric version_tuple to numpy.

Carreau opened the pull request:
This makes it easier and less error prone to write conditional code that
depends on numpy version.

We limit version tuple to be numeric only, and only major/minor/patch.

Dependencies can now check numpy version simply with:

    >>> import numpy
    ... assert numpy.version.version_tuple > (1, 22)

Without relying on more complex dependencies like
`packaging.version.parse`, or distutils' LooseVersion (which is now
deprecated).

This also prevent roll-out of custom version parsing code that might be
error prone, or worse, comparison of strings.

-- 

Note that I would call it `version_info`, as that the name IPython uses to expose it, but ~~matplotlib exposes `version_tuple`~~, and @rgommers seem to prefer `version_tuple`.

Note: matplotlib changed the version_tuple to `__version_info__`
<!--         ----------------------------------------------------------------
                MAKE SURE YOUR PR GETS THE ATTENTION IT DESERVES!
                ----------------------------------------------------------------

*  FORMAT IT RIGHT:
      http://www.numpy.org/devdocs/dev/development_workflow.html#writing-the-commit-message

*  IF IT'S A NEW FEATURE OR API CHANGE, TEST THE WATERS:
      http://www.numpy.org/devdocs/dev/development_workflow.html#get-the-mailing-list-s-opinion

*  HIT ALL THE GUIDELINES:
      https://numpy.org/devdocs/dev/index.html#guidelines

*  WHAT TO DO IF WE HAVEN'T GOTTEN BACK TO YOU:
      http://www.numpy.org/devdocs/dev/development_workflow.html#getting-your-pr-reviewed
-->

Carreau committed:
ENH: Add a numeric __version_info__ to numpy.

This makes it easier and less error prone to write conditional code that
depends on numpy version.

We limit version tuple to be numeric only, and only major/minor/patch.

Dependencies can now check numpy version simply with:

>>> import numpy
... assert numpy.version.__version_info__ > (1, 22)

Without relying on more complex dependencies like
`packaging.version.parse`, or distutils' LooseVersion (which is now
deprecated).

This also prevent roll-out of custom version parsing code that might be
error prone, or worse, comparison of strings.

We do not go as far as Matplotlib that use a named tuple and include
a/b/rc in the fourth component of the name tuple, but this could always
be added latter.

Co-authored-by: Bas van Beek <43369155+BvB93@users.noreply.github.com>
Co-authored-by: Clément Robert <cr52@protonmail.com>

BvB93 commented on numpy/version.py:
```diff
@@ -12,4 +12,29 @@
 
-del get_versions, vinfo
+
+def make_version_tuple():
```
```suggestion
def make_version_tuple() -> tuple[int, int, int]:
```
Could you add a return type annotation for the function?
`np.version` is one of the rare few module that contains inline annotations, rather than a dedicated stub file (as it is a rather simple module, annotation wise).

charris commented:
Needs a release note.

neutrinoceros commented on numpy/version.py:
```diff
@@ -12,4 +12,30 @@
+    dependencies to have conditional code without having to parse our version
+    string.
+
+    1) It can be tricky as 1) you don't want to compare string. 2) it can be
+    tempted to just split on dot and map convert to int, but that will fail on
```
```suggestion
    It can be tricky as 1) you don't want to compare string. 2) it can be
    tempting to just split on dot and map convert to int, but that will fail on
```

Carreau commented:
> Needs a release note.

Added.

Carreau commented:
A Nep would be fairly more complex IMHO. But if NumPy want to set the tone I'm happy to update IPython to follow.

seberg commented:
I somewhat like to include tuple in the name, but honestly: I don't care :).  NEP seems a bit much, but if anyone really wants to it...
If we go for a module global `__version_tuple__`, or whatever, a [SPEC](https://scientific-python.org/specs/) outlining all globals that packages should/could provide.  But IMO, that probably only gets interesting if someone proposes e.g. a standardized `__citation__` or so.

seberg commented:
Ah, one interesting point of reference:
```
>>> import sys
>>> sys.version_info
sys.version_info(major=3, minor=9, micro=10, releaselevel='final', serial=0)
```
Which is a named tuple.

neutrinoceros commented:
I agree that using a named-tuple would make sense since it's what the stdlib does. I would further advocate that using the same name as the standard library (`<x>.version_info`) would make these attributes more discoverable and less error-prone. It would also avoid patching IPython (and matplotlib is sorta close enough already)

mattip commented:
Has there been scientific-python wide discussion of package `version_info` and reusing the `sys.version_info` format?

rossbar commented:
> Has there been scientific-python wide discussion of package version_info and reusing the sys.version_info format?

I didn't see any related discussion in a quick search of [discuss.scientific-python.org](https://discuss.scientific-python.org/c/contributor/12) - IMO it's probably worth bringing up there, though doing isn't necessarily a blocker for this PR.

Carreau commented:
Bump on this, I cam across it recently again.

seberg commented:
I suppose even if we end up having both eventually it barely matters.  I agree it would be good to have something like a tuple, and while it would be nice to just agree generally on one thing, not sure it makes sense to keep waiting for that.
So unless, there is a clear push by anyone, maybe we should just put it in soon?

Carreau commented:
Hum , matplotlib seem to have changed from `version_tuple` to `__version_info__` (and made it a name tuple), so if you believe we should conform to what they are doing, I'm open to it.

As this is numpy and everybody copies you, I think you can set the consensus.

I'll followup to whatever is decided here and implement the pendant in IPython.

It is always possible to go tuple -> namedtuple and from 3-tuple to n-tuple later, so the only thing we should really worry about here is the naming.

Carreau commented:
I think I'm going to lean toward `__version_info__`, we could also decide to lean toward a NEP, or SPEC if you prefer to gather feedback.

seberg commented:
I like `__version_info__`, but not big opinion either way.  Since having something that is a tuple would always be useful.

mattip commented:
+1 for `__version_info__`

Carreau commented:
Rebase, squashed and updated to use `__version_info__`, I kept the three integer tuple for now as it's the lowest common denominator that can always be extended with more fields or upgraded to a name tuple later.

Another questions is : should this be exposed as `numpy.__version_info__` ?

seberg commented:
> Another questions is : should this be exposed as numpy.__version_info__ ?

That was my expectation, what do other projects do?

rgommers commented on numpy/version.py:
```diff
@@ -20,4 +20,34 @@
+
+    Matplotlib goes a bit further and make that a named tuple and include the
+    a/b/rc/... in the fourth place in the tuple, but this is generally
+    unnecessary. Though it could be added in a backward compatible manner.
```
Not doing this is a bit questionable, and makes it fairly fragile to work with development versions. Is there a reason not to do what Matplotlib does here? So for release versions it'd be a length-3 tuple, and otherwise a length-4 tuple?

Carreau replied:
You can't make it "just" a 4-tuple because:

```
>>> list(sorted(['rc', 'b', 'a', 'dev']))
['a', 'b', 'dev', 'rc']
```

And dev should be the highest. So you get the wrong order just with a tuple, and  you do need a custom object (like NumPyVersion, which I'm fine using). 

IMHO conditional checking of version number that contain a/b/rc/dev is an anti pattern, but I can see why you want it.


I don't like seeing 

```
np.__version_info__ < '1.22.0'
```

Because from an outside reader you don't know that `__version_info__` is a custom object that will parse `'1.22.0'` in its `__lt__`/`__gt__`/`__eq__`. And it makes it _easy_ for other project, or dev to think "oh it's ok to compare a string", and copy the bad practice. And the problem is that a lot of folks take numpy as example. 

Now my only goal here is to get a standard, hard to get wrong way of comparing versions, I'd like something that does not require super complex dependency or code (otherwise we should just rely on `from packaging.version import parse, Version` maybe? And a (int, int, int) seem like the simplest. I'm happy with more complicated though.

rgommers replied:
> IMHO conditional checking of version number that contain a/b/rc/dev is an anti pattern, but I can see why you want it.

I am more worried about `__version_info__ < (1, 2, 0)` simply being incorrect for any `1.2.0.dev0` version.

rgommers reviewed (commented):
> Another questions is : should this be exposed as `numpy.__version_info__` ?

~+1 from me to doing this.~ EDIT: after thinking about this some more and some recent discussions about `__version__` not being standard either (and some packaging folks actually recommending against adding it, for very much incorrect reasons), I am having second thoughts here. Comparing with `__version__` is always correct if done right, while this new `__version_info__` tuple won't be. Example using our public [NumPyVersion](https://numpy.org/devdocs/reference/generated/numpy.lib.NumpyVersion.html#numpy-lib-numpyversion):

```python
>>> ver = np.lib.NumpyVersion(np.__version__)
>>> ver <= '1.23.3.dev0'
False
```

So I am wondering now if we shouldn't just reify `__version__` and make using PEP 440 compliant version strings as easy as possible. It's not as easy as a numeric tuple, but on the other hand using the numeric tuple is taking a correctness shortcut that is going to be annoying when people use development versions on larger projects (it will likely result in some spurious bug reports because the version check relies on a feature that's actually not present yet).

So a possible alternative here is to use:
```python
__version_info__ = np.lib.NumPyVersion(np.__version__)  # with a non-numpy-specific repr, so it's directly usable for other projects too.
``` 
This would allow using it as:
```python
np.__version_info__ < '1.22.0'
```
which is just as usable as the tuple version.

> Since `numpy` (like `scipy` and `astropy`) already has a `.version` modele, another option would be `version.info`.

The `version` namespace was an unfortunately side-effect of generating version info at build time and not caring about public/private API splits I believe. A full namespace just for version info seems too much. I'd much prefer to get rid of that over time, and prefer `__version__` (and maybe `__version_info__`; since using things like `np.version.major` separately is going to have the same correctness issues as `__version_info__`).

rgommers commented:
> It is also sort of annoying that `importlib.metadata.version('<module>')` return a str and not a version object...

Yep - I'm not a fan of that API, because it doesn't look at the actual package and `__version__` but at the package metadata (which can very easily be wrong, especially in development setups).

> Not doing this is a bit questionable, and makes it fairly fragile to work with development versions. Is there a reason not to do what Matplotlib does here? So for release versions it'd be a length-3 tuple, and otherwise a length-4 tuple?

`packaging` is a pretty annoying dependency. I think https://pypi.org/project/pep440/ works best. And as a larger library, vendoring that as a single file is kinda nice (IIRC we had to deviate from the original code here in SciPy; making it designed to be single-file-vendorable would be useful). 

Taking a step back, I think what we want to recommend is:
1. add `__version__` with a PEP 440 compliant version string in every package
2. do all version comparisons with a simple API that always does the right thing, and only accepts PEP 440 compliant versions. I'd argue for that to be an updated version of `pep440`. I think we unfortunately cannot get away with plain tuple-of-ints without running into correctness problems.

mhvk commented:
@Carreau - indeed, you are right, splitting is not good enough, comparing strings directly is better.

Thinking more, I guess @rgommers's comment led me to the same conclusion he just posted: that ensure `__version__` is properly standardized.

Carreau commented:
> packaging is a pretty annoying dependency. I think https://pypi.org/project/pep440/ works best. And as a larger library, vendoring that as a single file is kinda nice (IIRC we had to deviate from the original code here in SciPy; making it designed to be single-file-vendorable would be useful).


Sure let me know if there is any changes needed, I'm happy to integrate them in the library. 

I can also try to implement a version object that properly compare.
